components/causal_ts/src/tso.rs

```diff
@@ -62,55 +69,205 @@ impl TsoBatch {
+    }
+
+    /// The excluded end timestamp after the last in batch.
+    pub fn excluded_end(&self) -> TimeStamp {
+        TimeStamp::compose(self.physical, self.logical_end)
+    }
+}
+
+/// `TsoBatchList` is a ordered list of `TsoBatch`. It aims to:
+/// 1. Cache more number of TSO to improve high availability. See issue #12794.
+///    `TsoBatch` can only cache at most 262144 TSO as logical clock is 18 bits.
+/// 2. Fully utilize cached TSO when some regions require latest TSO (e.g. in the scenario of leader transfer).
+///    Other regions without the requirement can still use older TSO cache.
+struct TsoBatchList {
+    inner: RwLock<TsoBatchListInner>,
+
+    /// Number of available TSO.
+    /// Using signed integer for avoiding a wrap around huge value as it's not precisely counted.
+    tso_count: AtomicI32,
+
+    /// Statistics of TSO usage.
+    tso_usage: AtomicU32,
+
+    /// Length of batch list. It is used to limit size for efficiency, and keep batches fresh.
+    capacity: u32,
+}
+
+// Inner data structure of batch list.
+// The reasons why `crossbeam_skiplist::SkipMap` is not chosen:
+// 1. In `flush()` procedure, a reader of `SkipMap` can still acquire a batch after the it is removed,
+//    which would violate the causality requirement. The `RwLock<BTreeMap>` avoid this scenario
+//    by lock synchronization.
+// 2. It is a scenario with much more reads than writes. The `RwLock` would be no less efficient
+//    than lock free implementation.
+type TsoBatchListInner = BTreeMap<u64, TsoBatch>;
+
+enum TsoBatchListIter<'a> {
+    Iter(btree_map::Iter<'a, u64, TsoBatch>),
+    Range(btree_map::Range<'a, u64, TsoBatch>),
+}
+
+impl<'a> TsoBatchListIter<'a> {
+    pub fn next(&mut self) -> Option<(&'a u64, &'a TsoBatch)> {
+        match self {
+            TsoBatchListIter::Iter(ref mut iter) => iter.next(),
+            TsoBatchListIter::Range(ref mut range) => range.next(),
+        }
+    }
+}
+
+impl TsoBatchList {
+    pub fn new(capacity: u32) -> Self {
+        Self {
+            inner: Default::default(),
+            tso_count: Default::default(),
+            tso_usage: Default::default(),
+            capacity,
+        }
+    }
+
+    pub fn tso_count(&self) -> u32 {
+        std::cmp::max(self.tso_count.load(Ordering::Relaxed), 0) as u32
+    }
+
+    pub fn tso_usage(&self) -> u32 {
+        self.tso_usage.load(Ordering::Relaxed)
+    }
 
-        if physical < self.physical
-            || (physical == self.physical && logical_start < self.logical_end)
+    pub fn take_and_report_tso_usage(&self) -> u32 {
+        let usage = self.tso_usage.swap(0, Ordering::Relaxed);
+        TS_PROVIDER_TSO_BATCH_LIST_COUNTING
+            .with_label_values(&["tso_usage"])
+            .observe(usage as f64);
+        usage
+    }
+
+    // TODO: make async
+    fn remove_batch(&self, key: u64) {
+        if let Some(batch) = self.inner.write().remove(&key) {
```

Comment:
This is a potential bottleneck here, since the lock is a sync op. Should we add a metrics to track down the time to remove batch?

Reply:
`remove_batch` will be changed to an asynchronous method in next PR.